topgun/k8s/dns_proxy_test.go

```diff
@@ -52,51 +55,61 @@ var _ = Describe("DNS Resolution", func() {
-			<-sess.Exited
-
-			if !c.shouldWork {
-				Expect(sess.ExitCode()).ToNot(BeZero())
-				return
-			}
-
-			Expect(sess.ExitCode()).To(BeZero())
-		},
-		Entry("Proxy Enabled, with full service name", Case{
-			enableDnsProxy:  "true",
-			addressFunction: fullAddress,
-			shouldWork:      true,
-		}),
-		Entry("Proxy Enabled, with short service name", Case{
-			enableDnsProxy:  "true",
-			addressFunction: shortAddress,
-			shouldWork:      false,
-		}),
-		Entry("Proxy Disabled, with full service name", Case{
-			enableDnsProxy:  "false",
-			addressFunction: fullAddress,
-			shouldWork:      true,
-		}),
-		Entry("Proxy Disabled, with short service name", Case{
-			enableDnsProxy:  "false",
-			addressFunction: shortAddress,
-			shouldWork:      true,
-		}),
-		Entry("Adding extra dns server, with Proxy Disabled and full address", Case{
-			enableDnsProxy:  "false",
-			dnsServer:       "8.8.8.8",
-			addressFunction: fullAddress,
-			shouldWork:      false,
-		}),
-		Entry("Adding extra dns server, with Proxy Enabled and full address", Case{
-			enableDnsProxy:  "true",
-			dnsServer:       "8.8.8.8",
-			addressFunction: fullAddress,
-			shouldWork:      false,
-		}),
-	)
+	expectedDnsProxyBehaviour := func(useContainerd bool) {
+		DescribeTable("different proxy settings",
+			func(c Case) {
+				setupDeployment(useContainerd, c.enableDnsProxy, c.dnsServer)
+
+				sess := fly.Start("execute", "-c", "tasks/dns-proxy-task.yml", "-v", "url="+c.addressFunction())
+				<-sess.Exited
+
+				if !c.shouldWork {
+					Expect(sess.ExitCode()).ToNot(BeZero())
+					return
+				}
+
+				Expect(sess.ExitCode()).To(BeZero())
+			},
+			Entry("Proxy Enabled, with full service name", Case{
+				enableDnsProxy:  "true",
+				addressFunction: fullAddress,
+				shouldWork:      true,
+			}),
+			Entry("Proxy Enabled, with short service name", Case{
+				enableDnsProxy:  "true",
+				addressFunction: shortAddress,
+				shouldWork:      false,
+			}),
+			Entry("Proxy Disabled, with full service name", Case{
+				enableDnsProxy:  "false",
+				addressFunction: fullAddress,
+				shouldWork:      true,
+			}),
+			Entry("Proxy Disabled, with short service name", Case{
+				enableDnsProxy:  "false",
+				addressFunction: shortAddress,
+				shouldWork:      true,
+			}),
+			Entry("Adding extra dns server, with Proxy Disabled and full address", Case{
+				enableDnsProxy:  "false",
+				dnsServer:       "8.8.8.8",
+				addressFunction: fullAddress,
+				shouldWork:      false,
+			}),
+			Entry("Adding extra dns server, with Proxy Enabled and full address", Case{
+				enableDnsProxy:  "true",
+				dnsServer:       "8.8.8.8",
+				addressFunction: fullAddress,
+				shouldWork:      false,
+			}),
+		)
+	}
+
+	Context("with gdn backend", func() {
+		expectedDnsProxyBehaviour(false)
+	})
+
+	Context("with containerd backend", func() {
+		expectedDnsProxyBehaviour(true)
+	})
```

Comment:
nit: could we move these up to the top so this test suite follows the flow of:

* BeforeEach()
* AfterEach()
* Context()